### PR TITLE
Add CLI-Flag to enable only specific collectors

### PR DESCRIPTION
**Pull request by ThomasBuchinger**
Hi,
I would like to have the possibility to select which of the collectors should be enabled. So I added a bit of meta-programming to create the Collector objects based on  CLI arguments.

By default it uses the same hardcoded collectors list, only with strings instead of the real objects.

btw. I am not happy with the place for `collector_names`. I don't want to put it in the parse_options method, as this part is already pretty dense, but I dislike having it outside the method, where it is used, too any ideas appriciated

### CODE REVIEW

**Comment by ThomasBuchinger** [comment]
amazingly the code works on my machine, when it should not.... I'll try to fix it and add the unit tests. 
Unfortunately the current version does not run in my environment. I have to work backwards from my develop version :(

**Comment by viennaa** [comment]
For your testing try running it in the docker container. Makefile is included in the root of the repo.

https://github.com/sapcc/vrops-exporter/blob/master/Makefile

**Comment by viennaa** [comment]
Please fix TestLaunchExporter test:
```======================================================================
ERROR: test_with_cli_and_env_params (__main__.TestLaunchExporter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/TestLaunchExporter.py", line 33, in test_with_cli_and_env_params
    parse_params()
  File "./exporter.py", line 35, in parse_params
    options.collectors = collector_names
NameError: name 'collector_names' is not defined

======================================================================
ERROR: test_with_cli_params_1 (__main__.TestLaunchExporter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/TestLaunchExporter.py", line 15, in test_with_cli_params_1
    parse_params()
  File "./exporter.py", line 35, in parse_params
    options.collectors = collector_names
NameError: name 'collector_names' is not defined

======================================================================
ERROR: test_with_cli_params_2 (__main__.TestLaunchExporter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/TestLaunchExporter.py", line 24, in test_with_cli_params_2
    parse_params()
  File "./exporter.py", line 35, in parse_params
    options.collectors = collector_names
NameError: name 'collector_names' is not defined

======================================================================
ERROR: test_without_params (__main__.TestLaunchExporter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/TestLaunchExporter.py", line 49, in test_without_params
    parse_params()
  File "./exporter.py", line 35, in parse_params
    options.collectors = collector_names
NameError: name 'collector_names' is not defined
```

**Comment by ThomasBuchinger** [comment]
@viennaa I think this one is ready for review from my side. Mind reviewing it?

File: exporter.py [hunk 1]
```diff
@@ -55,18 +65,18 @@ def run_prometheus_server(port, collectors,  *args):
...
-                DatastoreStatsCollector(),
-                VMStatsCollector(),
-                VMPropertiesCollector()
-                # add new collectors above this line
-                # CollectorUp()
-            ]
+
+    collectors = list(map(lambda c: initialize_collector_by_name(c), options.collectors))
+    if os.environ['DEBUG'] >= '1':
+        print("Started collectors: " + str(list(collectors)))
```
Comment (viennaa) [inline exporter.py:81]: Please don't dump objects here, name would be better to read.
```
Started collectors: [<collectors.HostSystemStatsCollector.HostSystemStatsCollector object at 0x7faa593f69d0>, <collectors.HostSystemPropertiesCollector.HostSystemPropertiesCollector object at 0x7faa5a7b0130>]
```

File: exporter.py [hunk 2]
```diff
@@ -55,18 +65,18 @@ def run_prometheus_server(port, collectors,  *args):
     while True:
         time.sleep(1)
 
+def initialize_collector_by_name(class_name):
+    class_module = importlib.import_module('collectors.%s' % (class_name))
```
Comment (viennaa) [inline exporter.py:69]: Please catch `ModuleNotFoundError:` in case there is no matching collector.

**Comment by ThomasBuchinger** [comment]
@viennaa  fixed the 2 changes and added a couple of unit tests

File: tests/TestLaunchExporter.py [hunk 3]
```diff
@@ -49,5 +69,26 @@ def test_without_params(self):
...
+        self.assertIsNone(collector)
+        self.assertEqual(mocked_print.mock_calls, [call('No Collector "BogusCollector" defined. Ignoring...')])
+
+    @patch('builtins.print')
+    def test_with_invalid_collector(self, mocked_print):
+        importlib.import_module = MagicMock(return_value=collectors.CollectorUp)
+        collector = initialize_collector_by_name('ClassNotDefinedCollector')
+        self.assertIsNone(collector)
+        self.assertEqual(mocked_print.mock_calls, [call('Unable to initialize "ClassNotDefinedCollector". Ignoring...')])
+
```
Comment (viennaa) [inline tests/TestLaunchExporter.py:92]: Good idea, please move to dedicated test file. Maybe take another collector, not sure if we are keeping the collectorUp-collector.

File: tests/TestLaunchExporter.py [hunk 4]
```diff
@@ -35,6 +37,24 @@ def test_with_cli_and_env_params(self):
...
+    def test_with_no_collector(self):
+        sys.argv = ['prog', '--port', '1234', '-i', 'inventory.some.url']
+        options = parse_params()
+        self.assertEqual(options.collectors, default_collectors(), 'Default collector list does not match the default')
+
+    # test with only one collector enabled
+    def test_with_one_collector(self):
+        sys.argv = ['prog', '--port', '1234', '-i', 'inventory.some.url', '-c', 'VMStatsCollector']
+        options = parse_params()
+        self.assertEqual(options.collectors, ['VMStatsCollector'], 'Default collector list does not match the default')
```
Comment (viennaa) [inline tests/TestLaunchExporter.py:50]: Please adjust message: `Collector list does not match given single collector.`

File: tests/TestLaunchExporter.py [hunk 5]
```diff
@@ -35,6 +37,24 @@ def test_with_cli_and_env_params(self):
...
+
+    # test multiple enabled collectors
+    def test_with_multiple_collector(self):
+        sys.argv = ['prog', '--port', '1234', '-i', 'inventory.some.url', '-c', 'VMStatsCollector', '-c', 'VMPropertiesCollector']
+        options = parse_params()
+        self.assertEqual(options.collectors, ['VMStatsCollector', 'VMPropertiesCollector'], 'Default collector list does not match the default')
```
Comment (viennaa) [inline tests/TestLaunchExporter.py:56]: Please adjust message: Collector list does not match given multiple collectors.

**Pull request merged**